### ISSUE TITLE
ir-remote.json Caused Bootloop

### ISSUE THREAD

**Issue by Crsarmv7l** [d03n3rfr1tz3/TTGO.T-Watch.2020#20]
Sorry to keep opening issues. 

I finally had some time to try out the multi-ir send and loaded up a test json with a remote repeated. The idea was to generate a bunch of flashes to ensure it worked prior to trying to introduce the delay we discussed in the other issue. Unfortunately after loading the json via ftp and rebooting, the ir-remote.json causes a bootloop.

I ran serial monitor and after reading ir-remote.json core 1 kernel panic'ed. The result is constant reboots after the init gui screen. I can post the full log if you would like.

As it never fully boots and I cannot access utilities to erase SPIFFS (and ir-remote.json), and it never boots fully to allow deletion of ir-remote.json via ftp.... how would I be able to delete SPIFFS via platformio or erase ir-remote.json?

**Comment by d03n3rfr1tz3** [comment]
You can just comment out the IR app in the main.cpp, so that it does not get loaded. The watch should then boot normally, if there is not another reason for a crash and therefore you should have plenty of time to erase or fix the ir-remote.json.

Just a guess, why it crashes: Depending on the actual crash reason, it might be possible that you have an error in your ir-remote.json. If you want, you can post your configuration and we both can have a look at it.

Of course it might also be possible that there is still a bug in the changed IR app code, but it worked at least with my ir-remote.json with about 70 buttons, where I changed a few of them to have multiple commands. If I did not do a stupid mistake copying over the tests to my example configuration (which still looks good to me), it should work at least with that example config. Only difference I see is, that I currently do not use raw-codes myself, but I'm pretty sure I had them in the config, when I tested it. Still, that could be, where a possible bug is.
https://github.com/d03n3rfr1tz3/TTGO.T-Watch.2020/blob/master/conf/ir-remote.json.example

**Comment by d03n3rfr1tz3** [comment]
Just tested it to make sure there is no bug in the code. I copied over the part with the raw-codes as well as the multi-commands from the linked example configuration above, which increased my IR app to 84 buttons and it worked perfectly. Also clicking on it worked, not just booting the watch. Now I'm a little bit more relaxed again 😊

**Comment by Crsarmv7l** [comment]
I would bet it is a problem with my ir-remote.json. I made a bunch of syntax errors with my last json.

Here is mine for reference. Wanted to start simple with the multi.
```c
{
   "pages": [{
            "Power": [{
                "m": 7,
                "hex": "E0E040BF"
            }, {
                "m": 7,
                "hex": "E0E040BF"
            }, {
                "m": 7,
                "hex": "E0E040BF"
            }, {
                "m": 7,
                "hex": "E0E040BF"
            }, {
                "m": 7,
                "hex": "E0E040BF"
            }, {
                "m": 7,
                "hex": "E0E040BF"
            }, {
                "m": 7,
                "hex": "E0E040BF"
            }, {
                "m": 7,
                "hex": "E0E040BF"
            }, {
                "m": 7,
                "hex": "E0E040BF"
            }, {
                "m": 7,
                "hex": "E0E040BF"
            }, {
                "m": 7,
                "hex": "E0E040BF"
            }, {
                "m": 7,
                "hex": "E0E040BF"
            }, {
                "m": 7,
                "hex": "E0E040BF"
            }, {
                "m": 7,
                "hex": "E0E040BF"
            }, {
                "m": 7,
                "hex": "E0E040BF"
            }, {
                "m": 7,
                "hex": "E0E040BF"
            }, {
                "m": 7,
                "hex": "E0E040BF"
            }, {
                "m": 7,
                "hex": "E0E040BF"
            }]
        } 
    ],
    "defBtnHeight": 190,
    "defBtnWidth": 195,
    "defSpacing": 1
}
```
Will definitely try commenting out the ir-remote and reflashing. If not I think I found a way to overwrite the existing json and if all else fails I think I have a way to erase spiff as well.

Thank you

**Comment by Crsarmv7l** [comment]
Commenting out the IR app worked great for bypassing the bootloop. Wish I had thought of that.

I will try working from your example json, but appreciate any feedback on my json posted above.

**Comment by Crsarmv7l** [comment]
So tried a few different things.

1. Straight example json. It loaded but there were no buttons or pages after "stop" so no multi-ir.
2. Removed all but multi-ir. That resulted in bootloop.

Can you post your known working json (at your convenience) for me to try?

**Comment by d03n3rfr1tz3** [comment]
Well, your json looks completly fine. That should work. The example config should also work, strange.

I am currently a little bit curtly and only on my mobile, but here are two configs that at least worked on my watch. The first is my everyday config, the second is extended with the raw codes and multi-commands. If they also crash on your side, we need to find the difference in the code.

`{
    "pages": [{
            "Power": {
                "m": 7,
                "hex": "E0E040BF"
            },
            "Source": {
                "m": 7,
                "hex": "E0E0807F"
            },
            "Guide": {
                "m": 7,
                "hex": "E0E0F20D"
            },
            "Ch+": {
                "m": 7,
                "hex": "E0E048B7"
            },
            "Up": {
                "m": 7,
                "hex": "E0E006F9"
            },
            "Ch-": {
                "m": 7,
                "hex": "E0E008F7"
            },
            "Left": {
                "m": 7,
                "hex": "E0E0A659"
            },
            "OK": {
                "m": 7,
                "hex": "E0E016E9"
            },
            "Right": {
                "m": 7,
                "hex": "E0E046B9"
            },
            "Back": {
                "m": 7,
                "hex": "E0E01AE5"
            },
            "Down": {
                "m": 7,
                "hex": "E0E08679"
            },
            "Exit": {
                "m": 7,
                "hex": "E0E0B44B"
            },
            "Play": {
                "m": 7,
                "hex": "E0E0E21D"
            },
            "Pause": {
                "m": 7,
                "hex": "E0E052AD"
            },
            "Stop": {
                "m": 7,
                "hex": "E0E0629D"
            }
        }, {
            "Power": {
                "m": 56,
                "hex": "0400E00FF"
            },
            "Disc": {
                "m": 56,
                "hex": "0400EB847"
            },
            "Eject": {
                "m": 56,
                "hex": "0400E807F"
            },
            "Rewind": {
                "m": 56,
                "hex": "0400EB04F"
            },
            "Up": {
                "m": 56,
                "hex": "0400E18E7"
            },
            "Forward": {
                "m": 56,
                "hex": "0400E8877"
            },
            "Left": {
                "m": 56,
                "hex": "0400ED827"
            },
            "OK": {
                "m": 56,
                "hex": "0400E38C7"
            },
            "Right": {
                "m": 56,
                "hex": "0400E58A7"
            },
            "Back": {
                "m": 56,
                "hex": "0400EE817"
            },
            "Down": {
                "m": 56,
                "hex": "0400E9867"
            },
            "Exit": {
                "m": 56,
                "hex": "0400ED42B"
            },
            "Play": {
                "m": 56,
                "hex": "0400E28D7"
            },
            "Pause": {
                "m": 56,
                "hex": "0400E4CB3"
            },
            "Stop": {
                "m": 56,
                "hex": "0400EC837"
            }
        }, {
            "Power": {
                "m": 3,
                "hex": "807F50AF"
            },
            "Optical": {
                "m": 3,
                "hex": "807F609F"
            },
            "Aux": {
                "m": 3,
                "hex": "807F40BF"
            },
            "Vol-": {
                "m": 3,
                "hex": "807F6A95"
            },
            "Mute": {
                "m": 3,
                "hex": "807FD02F"
            },
            "Vol+": {
                "m": 3,
                "hex": "807F7A85"
            }
        }, {
            "White": {
                "m": 3,
                "hex": "FFA857"
            },
            "Flash": {
                "m": 3,
                "hex": "FFB24D"
            },
            "On": {
                "m": 3,
                "hex": "FFB04F"
            },
            "Red": {
                "m": 3,
                "hex": "FF9867"
            },
            "Strobe": {
                "m": 3,
                "hex": "FF00FF"
            },
            "Off": {
                "m": 3,
                "hex": "FFF807"
            },
            "Green": {
                "m": 3,
                "hex": "FFD827"
            },
            "Fade": {
                "m": 3,
                "hex": "FF58A7"
            },
            "B+": {
                "m": 3,
                "hex": "FF906F"
            },
            "Blue": {
                "m": 3,
                "hex": "FF8877"
            },
            "Smooth": {
                "m": 3,
                "hex": "FF30CF"
            },
            "B-": {
                "m": 3,
                "hex": "FFB847"
            }
        }, {
            "White": {
                "m": 3,
                "hex": "FF22DD"
            },
            "Flash": {
                "m": 3,
                "hex": "FFD02F"
            },
            "Power": {
                "m": 3,
                "hex": "FF02FD"
            },
            "Red": {
                "m": 3,
                "hex": "FF1AE5"
            },
            "Jump": {
                "m": 3,
                "hex": "FFA05F"
            },
            "Play": {
                "m": 3,
                "hex": "FF827D"
            },
            "Green": {
                "m": 3,
                "hex": "FF9A65"
            },
            "Fade": {
                "m": 3,
                "hex": "FFE01F"
            },
            "B+": {
                "m": 3,
                "hex": "FF3AC5"
            },
            "Blue": {
                "m": 3,
                "hex": "FFA25D"
            },
            "Smooth": {
                "m": 3,
                "hex": "FFE01F"
            },
            "B-": {
                "m": 3,
                "hex": "FFBA45"
            }
        }
    ],
    "defBtnHeight": 33,
    "defBtnWidth": 65,
    "defSpacing": 2
}
`

`{
    "pages": [{
            "Power": {
                "m": 7,
                "hex": "E0E040BF"
            },
            "Source": {
                "m": 7,
                "hex": "E0E0807F"
            },
            "Guide": {
                "m": 7,
                "hex": "E0E0F20D"
            },
            "Ch+": {
                "m": 7,
                "hex": "E0E048B7"
            },
            "Up": {
                "m": 7,
                "hex": "E0E006F9"
            },
            "Ch-": {
                "m": 7,
                "hex": "E0E008F7"
            },
            "Left": {
                "m": 7,
                "hex": "E0E0A659"
            },
            "OK": {
                "m": 7,
                "hex": "E0E016E9"
            },
            "Right": {
                "m": 7,
                "hex": "E0E046B9"
            },
            "Back": {
                "m": 7,
                "hex": "E0E01AE5"
            },
            "Down": {
                "m": 7,
                "hex": "E0E08679"
            },
            "Exit": {
                "m": 7,
                "hex": "E0E0B44B"
            },
            "Play": {
                "m": 7,
                "hex": "E0E0E21D"
            },
            "Pause": {
                "m": 7,
                "hex": "E0E052AD"
            },
            "Stop": {
                "m": 7,
                "hex": "E0E0629D"
            }
        }, {
            "Power": {
                "m": 56,
                "hex": "0400E00FF"
            },
            "Disc": {
                "m": 56,
                "hex": "0400EB847"
            },
            "Eject": {
                "m": 56,
                "hex": "0400E807F"
            },
            "Rewind": {
                "m": 56,
                "hex": "0400EB04F"
            },
            "Up": {
                "m": 56,
                "hex": "0400E18E7"
            },
            "Forward": {
                "m": 56,
                "hex": "0400E8877"
            },
            "Left": {
                "m": 56,
                "hex": "0400ED827"
            },
            "OK": {
                "m": 56,
                "hex": "0400E38C7"
            },
            "Right": {
                "m": 56,
                "hex": "0400E58A7"
            },
            "Back": {
                "m": 56,
                "hex": "0400EE817"
            },
            "Down": {
                "m": 56,
                "hex": "0400E9867"
            },
            "Exit": {
                "m": 56,
                "hex": "0400ED42B"
            },
            "Play": {
                "m": 56,
                "hex": "0400E28D7"
            },
            "Pause": {
                "m": 56,
                "hex": "0400E4CB3"
            },
            "Stop": {
                "m": 56,
                "hex": "0400EC837"
            }
        }, {
            "Power": {
                "m": 3,
                "hex": "807F50AF"
            },
            "Optical": {
                "m": 3,
                "hex": "807F609F"
            },
            "Aux": {
                "m": 3,
                "hex": "807F40BF"
            },
            "Vol-": {
                "m": 3,
                "hex": "807F6A95"
            },
            "Mute": {
                "m": 3,
                "hex": "807FD02F"
            },
            "Vol+": {
                "m": 3,
                "hex": "807F7A85"
            }
        }, {
            "White": {
                "m": 3,
                "hex": "FFA857"
            },
            "Flash": {
                "m": 3,
                "hex": "FFB24D"
            },
            "On": {
                "m": 3,
                "hex": "FFB04F"
            },
            "Red": {
                "m": 3,
                "hex": "FF9867"
            },
            "Strobe": {
                "m": 3,
                "hex": "FF00FF"
            },
            "Off": {
                "m": 3,
                "hex": "FFF807"
            },
            "Green": {
                "m": 3,
                "hex": "FFD827"
            },
            "Fade": {
                "m": 3,
                "hex": "FF58A7"
            },
            "B+": {
                "m": 3,
                "hex": "FF906F"
            },
            "Blue": {
                "m": 3,
                "hex": "FF8877"
            },
            "Smooth": {
                "m": 3,
                "hex": "FF30CF"
            },
            "B-": {
                "m": 3,
                "hex": "FFB847"
            }
        }, {
            "White": {
                "m": 3,
                "hex": "FF22DD"
            },
            "Flash": {
                "m": 3,
                "hex": "FFD02F"
            },
            "Power": {
                "m": 3,
                "hex": "FF02FD"
            },
            "Red": {
                "m": 3,
                "hex": "FF1AE5"
            },
            "Jump": {
                "m": 3,
                "hex": "FFA05F"
            },
            "Play": {
                "m": 3,
                "hex": "FF827D"
            },
            "Green": {
                "m": 3,
                "hex": "FF9A65"
            },
            "Fade": {
                "m": 3,
                "hex": "FFE01F"
            },
            "B+": {
                "m": 3,
                "hex": "FF3AC5"
            },
            "Blue": {
                "m": 3,
                "hex": "FFA25D"
            },
            "Smooth": {
                "m": 3,
                "hex": "FFE01F"
            },
            "B-": {
                "m": 3,
                "hex": "FFBA45"
            } 
 ​        }, { 
 ​            "Home": { 
 ​                "m": 30,  ​                "raw": [3000,3000, 500,500, 500,1500, 500,1500, 500,500, 500,1500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,1500, 500,1500, 500,1500, 500,1500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,1500, 500,1500, 500,1500, 500,500, 500,1500, 500,1500, 500,1500, 500] 
 ​            }, 
 ​            "Up": { 
 ​                "m": 30,  ​                "raw": [3000,3000, 500,500, 500,1500, 500,1500, 500,500, 500,1500, 500,500, 500,500, 500,500, 500,1500, 500,500, 500,1500, 500,500, 500,500, 500,500, 500,1500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,1500, 500,500, 500,1500, 500,500, 500] 
 ​            }, 
 ​            "Auto": { 
 ​                "m": 30,  ​                "raw": [3000,3000, 500,500, 500,1500, 500,1500, 500,500, 500,1500, 500,500, 500,500, 500,500, 500,1500, 500,500, 500,1500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,1500, 500,500, 500,500, 500,500, 500] 
 ​            }, 
 ​            "Left": { 
 ​                "m": 30,  ​                "raw": [3000,3000, 500,500, 500,1500, 500,1500, 500,500, 500,1500, 500,500, 500,500, 500,500, 500,1500, 500,1500, 500,500, 500,500, 500,500, 500,1500, 500,1500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,1500, 500,500, 500,1500, 500,1500, 500,1500, 500,500, 500] 
 ​            }, 
 ​            "Pause": { 
 ​                "m": 30,  ​                "raw": [3000,3000, 500,500, 500,1500, 500,1500, 500,500, 500,1500, 500,500, 500,500, 500,500, 500,1500, 500,1500, 500,1500, 500,1500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,1500, 500,500, 500,1500, 500,1500, 500,500, 500,500, 500,500, 500] 
 ​            }, 
 ​            "Right": { 
 ​                "m": 30,  ​                "raw": [3000,3000, 500,500, 500,1500, 500,1500, 500,500, 500,1500, 500,500, 500,500, 500,500, 500,1500, 500,1500, 500,500, 500,1500, 500,500, 500,1500, 500,500, 500,1500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,1500, 500,1500, 500,1500, 500,1500, 500,500, 500,1500, 500] 
 ​            }, 
 ​            "Play": { 
 ​                "m": 30,  ​                "raw": [3000,3000, 500,500, 500,1500, 500,1500, 500,500, 500,1500, 500,500, 500,500, 500,500, 500,1500, 500,1500, 500,1500, 500,1500, 500,500, 500,1500, 500,500, 500,1500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,1500, 500,500, 500,1500, 500,1500, 500,1500, 500,500, 500,1500, 500] 
 ​            }, 
 ​            "Down": { 
 ​                "m": 30,  ​                "raw": [3000,3000, 500,500, 500,1500, 500,1500, 500,500, 500,1500, 500,500, 500,500, 500,500, 500,1500, 500,500, 500,1500, 500,1500, 500,500, 500,500, 500,1500, 500,1500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,1500, 500,1500, 500,500, 500,1500, 500,1500, 500] 
 ​            }, 
 ​            "Room": { 
 ​                "m": 30,  ​                "raw": [3000,3000, 500,500, 500,1500, 500,1500, 500,500, 500,1500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,1500, 500,1500, 500,500, 500,1500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,500, 500,1500, 500,500, 500,500, 500,1500, 500,1500, 500,1500, 500,500, 500,500, 500] 
 ​            } 
 ​        }, { 
 ​            "MultiPower": [{ 
 ​                "m": 7, 
 ​                "hex": "E0E040BF" 
 ​            }, { 
 ​                "m": 56, 
 ​                "hex": "0400E00FF" 
 ​            }, { 
 ​                "m": 3, 
 ​                "hex": "807F50AF" 
 ​            }], 
 ​            "MultiPlay": [{ 
 ​                "m": 7, 
 ​                "hex": "E0E0E21D" 
 ​            }, { 
 ​                "m": 56, 
 ​                "hex": "E0E0E21D" 
 ​            }], 
 ​            "MultiStop": [{ 
 ​                "m": 7, 
 ​                "hex": "0400E28D7" 
 ​            }, { 
 ​                "m": 56, 
 ​                "hex": "0400EC837" 
 ​            }] 
 ​        }
    ],
    "defBtnHeight": 33,
    "defBtnWidth": 65,
    "defSpacing": 2
}
`

**Comment by Crsarmv7l** [comment]
Thanks, I will give them a try in the next few days. I can only test when my daughter is napping.

If there are any crashes (hopefully not) I would be happy to provide any logs/code needed to troubleshoot.

During my testing I made sure I was using the current sharandac, with only the IRremote app replaced with your code. I will also try with your complete code as well, prior to updating here.

**Comment by Crsarmv7l** [comment]
Alright my friend, I have some testing results:

To make things specific the following were my steps:
1. Git cloned your master repo
2. Added lvgl 7.11 (7.7.2 functions the same in all my tests) as specified, due to this issue https://github.com/sharandac/My-TTGO-Watch/issues/319#issuecomment-991878952
3. Commented out all apps but Activity, Alarm, Calc, gps_status, IRController, Kodi_remote, and stopwatch in main.cpp
4. Deleted unused apps in src/app
5. Built the project
6. Uploaded the project to my T-watch 2020 v1
7. Uploaded example.json
8. Rebooted
9. Tested IR app
10. FTPed in and replaced the example.json with your Second remote config.
11. Rebooted
12. Tested the IR app.

In both cases the screen looks like the attached
![IRapp](https://user-images.githubusercontent.com/85343771/147398199-f59a1adf-df50-4606-a3ca-aedaa172eb5a.jpg)

No amount of swiping reveals any further buttons despite both configs having more buttons.

A config with only multi-remotes still bootloops.

**Comment by Crsarmv7l** [comment]
I was finally able to get your Second remote to load, display all pages, and verified multi-remote works.

That means something must be off with my json that is causing the boot loop, or the IRController app doesn't expect to see the double open [ right at the start.

**Comment by d03n3rfr1tz3** [comment]
Hey, glad to see that it somehow worked with one of my json configurations. I also reformatted them, so that they are easier to read.

You should be able to just use sharandac's master branch. He already merged the IR app update from the experimental branch into master just about a day after my PR. No need to fiddle with my specialized branch. 😅

I'll probably try your configuration later and have a closer look, if it also crashes on my side and why. I still cannot see any problem in your json, but there is something suspicious. Your configuration is very short and my configuration is very long. To successfully load my configuration, the initialization needs to reallocate a bigger button array. Maybe on shorter configurations something is broken with the initial allocated array size.
Another thing that strikes me, is your button size. Mine are 65x33 and your is 195x190. Shouldn't be the problem, as it should fit the display and would surprise me, if it crashes because of that, but you never know. We'll see.

**Comment by d03n3rfr1tz3** [comment]
Tested your JSON and it crashed exactly as you described. The position where it crashed was suspicious and matched my guess, that it has to do with your shorter configuration. Turns out, that the reallocation for longer configurations somehow fixes a problem with the initial size of the buttons array. I cannot really explain, why the initial size was wrong and why even moving it to the constructor did not work, but moving the allocation to the point where it loads the JSON next to the reallocation for bigger configurations, fixed the problem: https://github.com/d03n3rfr1tz3/TTGO.T-Watch.2020/commit/26634099d305e328ef0337782e58614e626409ae

After the fix I could load and run your posted configuration from above successfully. I already created a PR to sharandac's experimental branch. Good thing is, it reduces PSRAM usage a little bit, if the ir-remote.json is still empty or does not exist.
